proposals/p2188.md

```diff
@@ -0,0 +1,655 @@
+library evolution easier.
+
+We could always allow a struct pattern to match a struct with more fields,
+without requiring a `, _` suffix. This would aid evolution by reducing the cases
+where adding a field to a struct can be a breaking change, but such cases would
+still exist. Further, this would make matches against a struct-valued expression
+inconsistent with matches against a struct pattern.
+
+We could use a different syntax instead of `, _`. Other options that were
+explicitly considered:
+
+-   `, ...` seems visually evocative of "and more stuff", but risks conflicting
+    with variadic syntax or at least being confusing when used in a variadic
+    context, given that variadics are expected to claim `...` for pack
+    expansion.
+-   `, ._` suggests matching a field without specifying a name, but might create
+    an impression of matching just one field, and three low punctuation
+    characters in a row seems to be pushing the limits of readability.
+
+On balance, `, _` harmonizes well with the use of `_` to introduce a wildcard,
+without being too visually confusing given that the other use of `_` has a
+following `:`.
+
+We could remove the `{field: type}` shorthand and require
+`{.field = field: type}`. This may avoid encouraging reusing the field name even
+when it is not appropriate, and would remove some syntactic sugar that's not
+formally necessary. However, we expect this to be a common case whose ergonomics
+are important.
+
+We could use a different syntax for `{field: type}` that is less divergent from
+other struct syntaxes:
+
+-   `{.field: type}` seems like a contender, but doesn't work because that is
+    already recognized as a struct type literal. Also, the use of normal binding
+    syntax means that every locally-introduced name is always introduced by
+    `name: type` where `name` is not preceded by `.`.
+-   `{.=field: type}` might be a reasonable mnemonic shorthand for
+    `{.field = field: type}`, but looks a little surprising. This is probably
+    the best choice if concerns are found with `{field: type}` syntax.
+
+### Type pattern matching
+
+We could treat type deduction as a form of pattern matching. For example, we
+could allow
+
+```
+fn F(a: Vector(T:! Type)) -> T { return a[0]; }
+fn G() -> i32 {
+  let v: Vector(i32) = (1, 2, 3);
+  // Deduces `T = i32`.
+  return F(v);
+}
+```
+
+where the value of `T` is determined by pattern-matching `Vector(T:! Type)`
+against the supplied type `Vector(i32)` of `v`. And symmetrically:
+
+```
+fn H[m: i32, n: i32](k: i32, (m, n)) { return m + n; }
+fn I() {
+  // Deduces `m = 2`, `n = 3`.
+  H(1, (2, 3));
+}
+```
+
+This would ensure consistency between pattern matching and deduction,
+potentially reducing the number of rules that Carbon developers need to learn.
+
+We find that attempting to unify pattern-matching and type deduction in this way
+harms readability. Having distinct syntax for type-level matching and
```

Comment:
I don't know that I disagree that the unification harms readability, but your examples are contrived so it's really hard to imagine how this would impact typical code.